src/libostree/ostree-bootloader-zipl.c

```diff
@@ -78,8 +85,197 @@ _ostree_bootloader_zipl_write_config (OstreeBootloader  *bootloader,
+  *options = g_strdup(val);
+
+  return TRUE;
+}
+
+static gboolean
+_ostree_secure_execution_luks_key_exists (void)
+{
+  return (access(SECURE_EXECUTION_LUKS_ROOT_KEY, F_OK) == 0 &&
+          access(SECURE_EXECUTION_LUKS_CONFIG, F_OK) == 0);
+}
+
+static gboolean
+_ostree_secure_execution_enable_luks(gchar* initram, GError **error)
+{
+  static const char *cmd = "/usr/libexec/libostree/s390x-se-luks-gencpio";
+  const char *const argv[] = {cmd, initram, SECURE_EXECUTION_INITRD_IMAGE, NULL};
+
+  g_autofree gchar *out = NULL;
+  g_autofree gchar *err = NULL;
+  int status = 0;
+  if (!g_spawn_sync (NULL, (char**)argv, NULL, G_SPAWN_SEARCH_PATH,
+                     NULL, NULL, &out, &err, &status, error))
+    return glnx_prefix_error(error, "s390x SE: spawning %s", cmd);
+
+  if (!g_spawn_check_exit_status (status, error))
+    {
+      g_printerr("s390x SE: `%s` stdout: %s\n", cmd, out);
+      g_printerr("s390x SE: `%s` stderr: %s\n", cmd, err);
+      return glnx_prefix_error(error, "s390x SE: `%s` failed", cmd);
+    }
+
+  sd_journal_print(LOG_INFO, "s390x SE: luks key added to initrd");
+  return TRUE;
+}
+
+static gboolean
+_ostree_secure_execution_generate_sdboot (gchar *vmlinuz,
+                                          gchar *initram,
+                                          gchar *options,
+                                          GPtrArray *keys,
+                                          GError **error)
+{
+  if (!vmlinuz || !initram || !options || !keys)
+    return glnx_throw (error, "s390x SE: vmlinuz, initram, options, keys are missing");
+
+  g_autoptr(GString) cmd   = NULL;
+  g_autofree char *cmdline = g_strdup ("/tmp/sd_boot.parmfile.XXXXXX");
+  glnx_autofd int fd       = g_mkstemp (cmdline);
+
+  if (glnx_loop_write (fd, options, strlen (options)) < 0)
+    return glnx_throw_errno_prefix (error, "s390x SE: creating %s", cmdline);
+  glnx_close_fd(&fd);
+
+  cmd = g_string_new ("genprotimg");
+  g_string_append_printf (cmd, " -i %s", vmlinuz);
+
+  if (_ostree_secure_execution_luks_key_exists ())
+    {
+      if ( !_ostree_secure_execution_enable_luks (initram, error))
+        return FALSE;
+      g_string_append_printf (cmd, " -r %s", SECURE_EXECUTION_INITRD_IMAGE);
+    } else {
+      g_string_append_printf (cmd, " -r %s", initram);
+    }
+
+  g_string_append_printf (cmd, " -p %s", cmdline);
+
+  sd_journal_print(LOG_INFO, "s390x SE: kernel: %s", vmlinuz);
+  sd_journal_print(LOG_INFO, "s390x SE: initrd: %s", initram);
+  sd_journal_print(LOG_INFO, "s390x SE: kargs: %s", options);
+
+  for (guint i = 0; i < keys->len; ++i) {
+      gchar *key = g_ptr_array_index (keys, i);
+      g_string_append_printf (cmd, " -k %s", key);
+      sd_journal_print(LOG_INFO, "s390x SE: key[%d]: %s", i + 1, key);
+    }
+  g_string_append (cmd, " --no-verify");
+  g_string_append (cmd, " -o " SECURE_EXECUTION_BOOT_IMAGE);
+
+  gint status = 0;
+  if (!g_spawn_command_line_sync (cmd->str, NULL, NULL, &status, error))
```

Comment:
Hmm, why not use `g_spawn_sync` instead for this? You can use a `GPtrArray` to build up your argument list progressively.

Reply:
Agreed on this, passing the args directly will avoid shell quoting issues.